src/providers/postgres/qgspostgresfeatureiterator.cpp

```diff
@@ -431,7 +452,8 @@ bool QgsPostgresFeatureIterator::declareCursor( const QString& whereClause )
 
     // reloading the fields might help next time around
     // TODO how to cleanly force reload of fields?  P->loadFields();
-    close();
+    if ( closeOnFail )
+      close();
     return false;
```

Comment:
I think we should add something to the message log and the debug output if declareCursor fails.